lib/Sema/CSDiagnostics.h

```diff
@@ -479,16 +479,33 @@ class LabelingFailure final : public FailureDiagnostic {
 /// type without optional chaining or force-unwrapping it first.
 class MemberAccessOnOptionalBaseFailure final : public FailureDiagnostic {
   DeclNameRef Member;
+  Type MemberBaseType;
   bool ResultTypeIsOptional;
 
 public:
   MemberAccessOnOptionalBaseFailure(const Solution &solution,
                                     ConstraintLocator *locator,
-                                    DeclNameRef memberName, bool resultOptional)
+                                    DeclNameRef memberName,
+                                    Type memberBaseType,
+                                    bool resultOptional)
       : FailureDiagnostic(solution, locator), Member(memberName),
-        ResultTypeIsOptional(resultOptional) {}
+        MemberBaseType(memberBaseType), ResultTypeIsOptional(resultOptional) {}
 
   bool diagnoseAsError() override;
+  
+  Type getMemberBaseType() const {
+    return getSolution().simplifyType(MemberBaseType)->getRValueType();
+  }
+  
+  SourceLoc getLoc() const override {
+    if (getLocator()->isForKeyPathComponent()) {
+      return getSourceRange().End;
+    }
+    return FailureDiagnostic::getLoc();
```

Comment:
Do we need to special case key path components here? IMO it would make more sense if we returned `getSourceRange().End` for both cases here as then we'll always point to the dot in the member access.

Reply:
Nice, I just had this way because I thought that for non key path components the range was the member range, so it would point to the end of the member no the dot at the end of the base.

Reply:
Ah, I've adjusted that, seems like it is ok :)

Reply:
Great!